### ISSUE TITLE
Make it possible to customize generation config for Trainer's training loop evaluation

### ISSUE THREAD

**Issue by antonioalegria** [huggingface/transformers#24427]
### Feature request

When using `predict_with_generate` and we want to compute generation-based metrics during the eval happening during training, it would be good if the model's generation config is used and/or if we can pass the indented generation config into the train method so that it can be passed to evaluate.

As it is, the generation done is using the default parameters only.

### Motivation

The current way GenerationConfigs are used is pretty inconsistent and muddled IMO. You can set it at the model level but it's only used sometimes. You can pass it directly to evaluate, predict or generate but it's not clear if you should pass it as kwargs or as a full GenerationConfig.

Would be great to clean this up so that it's super clear on how to use it and have a very consistent way to use it, as in python. My suggestion would be to set it at the Trainer level and be able to override it in the evaluate, predict, generate methods with a simple generation_config: GenerationConfig parameter.

### Your contribution

Happy to discuss different possibilities and see where I could help.

**Comment by sgugger** [comment]
You can customize the model `generation_config` as you want and it will be the one being used. cc @gante to make sure I'm not saying something untrue.

**Comment by antonioalegria** [comment]
Maybe it's because I was setting the Generation Config in the model without putting _from_model_config = False and it was triggering the following, which resets the config:

```
if generation_config is None:
            # legacy: users may modify the model configuration to control generation -- update the generation config
            # model attribute accordingly, if it was created from the model config
            if self.generation_config._from_model_config:
                new_generation_config = GenerationConfig.from_model_config(self.config)
                if new_generation_config != self.generation_config:
                    warnings.warn(
                        "You have modified the pretrained model configuration to control generation. This is a"
                        " deprecated strategy to control generation and will be removed soon, in a future version."
                        " Please use a generation configuration file (see"
                        " https://huggingface.co/docs/transformers/main_classes/text_generation)"
                    )
                    self.generation_config = new_generation_config
            generation_config = self.generation_config
```

**Comment by antonioalegria** [comment]
Yeah, that was it. Though it is a bit confusing because that message says it's a deprecated strategy but doesn't say it will revert the configs.

So, maybe docs should be improved to say when Generation Config is used where, with what priority between the default one, the model one, the gen_kwargs, etc...

For example, in Se2SeqTrainer.evaluate and predict there is this code which doesn't take into consideration if there is a generation config set in the model.

```
gen_kwargs = gen_kwargs.copy()
if gen_kwargs.get("max_length") is None and gen_kwargs.get("max_new_tokens") is None:
    gen_kwargs["max_length"] = self.args.generation_max_length
gen_kwargs["num_beams"] = (
    gen_kwargs["num_beams"] if gen_kwargs.get("num_beams") is not None else self.args.generation_num_beams
)
```

If you do have a Generation Config set in the model or passed as generation_config parameter in evaluate/predict, where you set `max_new_tokens` this will yield a warning Both `max_new_tokens` and `max_length seem to have been set. It was the code above that set max_length because it didn't see the passed GenerationConfig.

Let me know if I should close this.

**Comment by antonioalegria** [comment]
Also, the shape[1] of outputs.predictions coming out of Seq2SeqTrainer.predict is 20 and it does not respect max_new_tokens passed in the generation config.

**Comment by gante** [comment]
Hi @antonioalegria 👋 

We do support parameterizing `Seq2SeqTrainer` with a `GenerationConfig` object. You seem to be hitting an issue due to `_from_model_config` being `True`, which simply means that you've issued a sequence of commands that we did not account for :)

Most of the issues you described are intentional temporary workarounds -- when a new feature is introduced, we have to ensure our library goes through a deprecation cycle, in which the old way of doing things take precedence. That's why you see strange patterns like the use of `_from_model_config` or "duplicated" pieces of code to control the same thing. Due to the large number of possibilities within `transformers`, sometimes we simply miss a few combinations.

That being said, let's start with the basics (which you haven't done 😉 ): what version of transformers are you using, and how can I reproduce your issue?

**Comment by antonioalegria** [comment]
Hi, I have removed the `_from_model_config` from the duplicated and altered config and the first issue no longer happens.

I still get those "Both max_new_tokens and max_length seem to have been set." warnings though.

transformers: 4.30.2

To reproduce use this code you just need to call `Seq2SeqTrainer.evaluate` with `generation_config=your_gen_config`, or set the generation_config in the model. In any case, you have to set max_new_tokens.

You will then see those warnings, which shouldn't happen.

Let me know if you'd like me to provide a running script.

**Comment by nick-maykr** [comment]
So what is the correct way to parametrize the generation (e.g. to use contrastive search) during the model training? [The documentation](https://huggingface.co/docs/transformers/main_classes/text_generation) misses this point.

**Comment by nick-maykr** [comment]
@gante @sgugger sorry to bother you, but it seems like the issue with GenerationConfig has not been resolved. Please, take a look at this code:
```
my_config = GenerationConfig.from_pretrained(CHECKPOINT_DIR)
my_config.forced_bos_token_id = tokenizer.encode("[")
my_config.forced_eos_token_id = tokenizer.encode("]")
my_config.max_new_tokens = 1000
my_config.max_length = 1000

args = Seq2SeqTrainingArguments(
    ....
    generation_config=my_config,
)

trainer = Seq2SeqTrainer(
    model_init=model_init,
    args=args,
    train_dataset=tokenized_dataset["train"],
    eval_dataset=tokenized_dataset["test"],
    data_collator=data_collator,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)
```

I check the config before starting the training, and everything looks fine:
```
>>> trainer.model.generation_config
GenerationConfig {
  "decoder_start_token_id": 0,
  "eos_token_id": 2,
  "forced_bos_token_id": [
    63
  ],
  "forced_eos_token_id": [
    65
  ],
  "max_length": 1000,
  "max_new_tokens": 1000,
  "pad_token_id": 0,
  "transformers_version": "4.32.1"
}
```
Then during the evaluation phase of the training I print the generated sentences and notice that the model doesn't follow my generation config. The forces start and end tokens are ignored and the number of generated tokens is 20 for any sample. I then stop the training and run the previous command again, and to my surprise see the following:
```
>>> trainer.model.generation_config
GenerationConfig {
  "decoder_start_token_id": 0,
  "eos_token_id": 2,
  "pad_token_id": 0,
  "transformers_version": "4.32.1"
}
``` 
Why has the config been reset and how can I avoid it? It seems like a bug to me.

@antonioalegria FYI

**Comment by gante** [comment]
Hey @nick-maykr 👋 Thank you for raising the issue!

I've just opened a PR that tackles this issue (fine-tuning from older models often ignoring `generation_config` changes), which should no longer happen after it gets merged 👉 https://github.com/huggingface/transformers/pull/25962

As for the `max_length` warning:
- The warning should go away if you unset `max_length` (I see `"max_length": 1000`). `max_length` and `max_new_tokens` are mutually exclusive.
- The value `20` in the warning is a bug, it should load and display the `1000` you set (I'll open a PR)

Let me know if you come across further issues. As I've written above, we have changed how we control generation late 2022, and maintaining retrocompatibility is challenging :)
